File: packages/dcflight/ios/Classes/channel/events/DCMauiLayoutMethodHandler.swift
/*
 * Copyright (c) Dotcorr Studio. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */


import Flutter
import UIKit

/// Handles layout method channel interactions between Flutter and native code
public class DCMauiLayoutMethodHandler: NSObject {
    /// Singleton instance
    public static let shared = DCMauiLayoutMethodHandler()
    let frame = UIScreen.main.bounds;
    
    /// Method channel for layout operations
    var methodChannel: FlutterMethodChannel?
    
    /// Initialize with Flutter binary messenger
    func initialize(with binaryMessenger: FlutterBinaryMessenger) {
        methodChannel = FlutterMethodChannel(
            name: "com.dcmaui.layout",
            binaryMessenger: binaryMessenger
        )
        
        methodChannel?.setMethodCallHandler(handleMethodCall)
        
    }
    
    /// Handle method calls from Flutter
    private func handleMethodCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getScreenDimensions":
            handleGetScreenDimensions(result: result)
            
        case "setUseWebDefaults":
            handleSetUseWebDefaults(call: call, result: result)
            
        case "setLayoutAnimationEnabled":
            handleSetLayoutAnimationEnabled(call: call, result: result)
            
        case "getLayoutAnimationEnabled":
            handleGetLayoutAnimationEnabled(result: result)
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func handleGetScreenDimensions(result: @escaping FlutterResult) {
        let bounds = UIScreen.main.bounds
        let dimensions = [
            "width": bounds.width,
            "height": bounds.height,
            "scale": UIScreen.main.scale,
            "statusBarHeight": UIApplication.shared.statusBarFrame.height
        ]
        
        result(dimensions)
    }
    
    private func handleSetLayoutAnimationEnabled(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let enabled = args["enabled"] as? Bool else {
            result(FlutterError(
                code: "ARGS_ERROR",
                message: "setLayoutAnimationEnabled requires 'enabled' boolean parameter",
                details: nil
            ))
            return
        }
        
        let duration = (args["duration"] as? NSNumber)?.doubleValue ?? 0.3
        
        DispatchQueue.main.async {
            DCFLayoutManager.shared.layoutAnimationEnabled = enabled
            if enabled {
                DCFLayoutManager.shared.layoutAnimationDuration = duration
            }
            result(true)
        }
    }
    
    private func handleGetLayoutAnimationEnabled(result: @escaping FlutterResult) {
        result([
            "enabled": DCFLayoutManager.shared.layoutAnimationEnabled,
            "duration": DCFLayoutManager.shared.layoutAnimationDuration
        ])
    }
    
    private func handleSetUseWebDefaults(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let enabled = args["enabled"] as? Bool else {
            result(FlutterError(code: "INVALID_ARGUMENTS", 
                              message: "setUseWebDefaults requires 'enabled' boolean parameter", 
                              details: nil))
            return
        }
        
        DCFLayoutManager.shared.setUseWebDefaults(enabled)
        result(true)
    }
}

File: packages/dcflight/android/src/main/kotlin/com/dotcorr/dcflight/bridge/DCMauiLayoutMethodHandler.kt
/*
 * Copyright (c) Dotcorr Studio. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.dotcorr.dcflight.bridge

import android.os.Handler
import android.os.Looper
import android.util.Log
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.Result
import com.dotcorr.dcflight.layout.YogaShadowTree
import com.dotcorr.dcflight.layout.DCFLayoutManager

/**
 * Handles layout-related method calls from Flutter
 */
class DCMauiLayoutMethodHandler : MethodChannel.MethodCallHandler {

    companion object {
        private const val TAG = "DCMauiLayoutMethodHandler"

        @JvmField
        val shared = DCMauiLayoutMethodHandler()

        fun initialize(binaryMessenger: io.flutter.plugin.common.BinaryMessenger) {
            val channel = MethodChannel(binaryMessenger, "com.dotcorr.dcflight/layout")
            channel.setMethodCallHandler(shared)
        }
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "calculateLayout" -> {
                handleCalculateLayout(call.arguments as? Map<String, Any>, result)
            }
            "setUseWebDefaults" -> {
                handleSetUseWebDefaults(call.arguments as? Map<String, Any>, result)
            }
            "getUseWebDefaults" -> {
                handleGetUseWebDefaults(result)
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    private fun handleCalculateLayout(args: Map<String, Any>?, result: Result) {
        val nodeId = args?.get("nodeId") as? String

        if (nodeId == null) {
            result.error("LAYOUT_ERROR", "Invalid node ID", null)
            return
        }

        Handler(Looper.getMainLooper()).post {
            val layout = YogaShadowTree.shared.calculateLayout(nodeId)
            if (layout != null) {
                result.success(layout)
            } else {
                result.error("LAYOUT_FAILED", "Failed to calculate layout for node $nodeId", null)
            }
        }
    }

    private fun handleSetUseWebDefaults(args: Map<String, Any>?, result: Result) {
        val enabled = args?.get("enabled") as? Boolean

        if (enabled == null) {
            result.error("ARGS_ERROR", "Invalid enabled parameter", null)
            return
        }

        Handler(Looper.getMainLooper()).post {
            DCFLayoutManager.shared.setUseWebDefaults(enabled)
            result.success(true)
        }
    }

    private fun handleGetUseWebDefaults(result: Result) {
        Handler(Looper.getMainLooper()).post {
            val enabled = DCFLayoutManager.shared.getUseWebDefaults()
            result.success(enabled)
        }
    }
}